hack/version_discovery.py

```diff
@@ -0,0 +1,250 @@
+                    if repo_results:
+                        results.extend(repo_results)
+                except Exception as e:
+                    logger.error(f"Error processing {repo}: {e}")
+
+        return results
+
+    def process_repository(
+        self, repo: str, config: RepositoryConfig
+    ) -> list[RepositoryInfo]:
+        version_prefix = config.get("version_prefix", "v")
+        images = config.get("images", [])
+
+        try:
+            version_info = self.find_latest_release_version(repo, version_prefix)
+            if version_info:
+                tag_name, _ = version_info
+                return [
+                    {
+                        "repository": f"https://github.com/{repo}",
+                        "ref": tag_name,
+                        "image_url": None,
+                    }
+                ]
+        except Exception as e:
+            logger.error(f"Error finding latest release for {repo}: {e}")
+
+        if not images:
+            logger.warning(
+                f"No releases or tags found for {repo} and no images configured"
+            )
+            return []
+
+        repo_results: list[RepositoryInfo] = []
+        for image in images:
+            component_name = f"{repo}/{image.split('/')[-1]}"
+            logger.info(f"Checking {component_name} latest commit")
+
+            try:
+                commit_info = self.find_latest_commit_with_image(repo, image)
+                if commit_info:
+                    commit_sha, tag = commit_info
+                    repo_results.append(
+                        {
+                            "repository": f"https://github.com/{repo}",
+                            "ref": commit_sha,
+                            "image_url": f"{image}:{tag}",
+                        }
+                    )
+                else:
+                    logger.warning(f"No images found for {component_name}")
+            except Exception as e:
+                logger.error(f"Error finding commit with image for {repo}/{image}: {e}")
+
+        return repo_results
+
+    @staticmethod
+    def save_to_release_candidates(
+        results: list[RepositoryInfo], filename: str = "release-candidates.yaml"
+    ) -> None:
+        repo = ReleaseCandidatesRepository(filename)
+        candidate = ReleaseCandidate(components=results)
+        if repo.save(candidate):
+            logger.info(
+                f"Saved new release candidate {candidate.metadata.get('id')} to {filename}"
+            )
+        else:
+            logger.info("No changes detected or error occurred, skipping update")
+
+
+def main() -> None:
+    try:
+        logger.info("Starting component scan")
+        scanner = ComponentScanner()
+        results = scanner.scan_components()
+        if not results:
+            logger.error("No components found")
+            sys.exit(1)
+
+        scanner.save_to_release_candidates(results)
```

Comment:
Let's try to follow the single responsibility principle: ComponentScanner should only scan. Then we can save its finding through another class, for example ReleaseCandidateRepository or some other abstraction on top of it.

I think at the moment the component scanner has too much knowledge about how we deal with release candidates (for example IMO it should not know the filename where we want to save to)